pkg/controller/gcp/storage/bucket.go

```diff
@@ -0,0 +1,315 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/apis/gcp/storage/v1alpha1"
+	gcpv1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/gcp/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/util"
+)
+
+const (
+	controllerName = "handle.storage.gcp.crossplane.io"
+	finalizer      = "finalizer." + controllerName
+
+	reconcileTimeout      = 1 * time.Minute
+	requeueAfterOnSuccess = 30 * time.Second
+
+	// failed condition reasons
+	//
+	// failed to get resource handler
+	failedToGetHandler = "error getting handler"
+	// failed to delete bucket resource
+	failedToDelete = "error deleting"
+	// failed to retrieve bucket resource
+	failedToRetrieve = "error retrieving"
+	// failed to create bucket resource
+	failedToCreate = "error creating"
+	// failed to update bucket resource
+	failedToUpdate = "error updating"
+	// failed to save connection secret
+	failedToSaveSecret = "error saving connection secret"
+)
+
+var (
+	resultRequeue    = reconcile.Result{Requeue: true}
+	requeueOnSuccess = reconcile.Result{RequeueAfter: requeueAfterOnSuccess}
+)
+
+// Reconciler reconciles a Provider obj
+type Reconciler struct {
+	client.Client
+	factory
+}
+
+// Add creates a newHandler Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	r := &Reconciler{
+		Client:  mgr.GetClient(),
+		factory: &bucketFactory{mgr.GetClient()},
+	}
+
+	// Create a newHandler controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Provider
+	err = c.Watch(&source.Kind{Type: &v1alpha1.Bucket{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the cluster for a Provider obj and makes changes based on the state read
+// and what is in the Provider.Spec
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("reconciling %s: %v", v1alpha1.BucketKindAPIVersion, request)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	b := &v1alpha1.Bucket{}
+	if err := r.Get(ctx, request.NamespacedName, b); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	bh, err := r.newHandler(ctx, b)
+	if err != nil {
+		b.Status.SetFailed(failedToGetHandler, err.Error())
+		return resultRequeue, r.Status().Update(ctx, b)
+	}
+
+	// Check for deletion
+	if b.DeletionTimestamp != nil {
+		return bh.delete(ctx)
+	}
+
+	return bh.sync(ctx)
+}
+
+type factory interface {
```

Comment:
Ah.. yeah.. I personally don't like word "factory", it is very... what's the word I am looking for... "Java'ish" like. Since I am so terrible at naming things, I will take ANY suggestion to replace this.